File: api/app/models/voucher.ts
import { DateTime } from 'luxon'
import { BaseModel, belongsTo, column } from '@adonisjs/lucid/orm'
import Student from './student.js'
import VoucherType from './voucher_type.js'
import VoucherSerie from './voucher_serie.js'
import PaymentType from './payment_type.js'
import type { BelongsTo } from '@adonisjs/lucid/types/relations'

export default class Voucher extends BaseModel {
  @column({ isPrimary: true })
  declare id: number

  @column()
  declare subtotal: number

  @column()
  declare taxValue: number

  @column()
  declare taxPercentage: number

  @column()
  declare total: number

  @column()
  declare change: number

  @column()
  declare set: string

  @column()
  declare correlative: string

  @column()
  declare studentId: number

  @column()
  declare voucherTypeId: number

  @column()
  declare voucherSeriesId: number

  @column()
  declare paymentTypeId: number

  @belongsTo(() => Student)
  declare student: BelongsTo<typeof Student>

  @belongsTo(() => VoucherType)
  declare voucherType: BelongsTo<typeof VoucherType>

  @belongsTo(() => VoucherSerie)
  declare voucherSerie: BelongsTo<typeof VoucherSerie>

  @belongsTo(() => PaymentType)
  declare paymentType: BelongsTo<typeof PaymentType>

  @column.dateTime({ autoCreate: true })
  declare createdAt: DateTime

  @column.dateTime({ autoCreate: true, autoUpdate: true })
  declare updatedAt: DateTime

  @column.dateTime()
  declare deletedAt: DateTime | null
}

File: api/src/main/kotlin/com/longdrink/app/model/Voucher.kt
package com.longdrink.app.model

import com.longdrink.app.common.Auditable
import jakarta.persistence.CascadeType
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.OneToMany
import jakarta.persistence.Table
import java.math.BigDecimal
import java.time.LocalDate
import java.time.LocalDateTime

@Entity
@Table(name = "vouchers")
data class Voucher(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long = 0,

    val subtotal: BigDecimal,

    @Column(name = "tax_value")
    val taxValue: BigDecimal,

    @Column(name = "tax_percentage")
    val taxPercentage: BigDecimal,

    val total: BigDecimal,

    val change: BigDecimal,

    val set: String = "",

    val correlative: String = "",

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "student_id")
    val student: Student,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "voucher_type_id", nullable = false)
    val voucherType: VoucherType,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "voucher_series_id", nullable = false)
    val voucherSeries: VoucherSeries,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "payment_type_id", nullable = false)
    val paymentType: PaymentType,

    @OneToMany(mappedBy = "voucher", cascade = [CascadeType.ALL], orphanRemoval = true)
    val details: MutableList<VoucherDetail> = mutableListOf(),

    @Column(name = "deleted_at")
    val deletedAt: LocalDateTime? = null,
) : Auditable()